pippy/PipelineStage.py

```diff
@@ -485,6 +486,36 @@ def _send_grads(
         return grad_send_reqs
 
     def forward(self, *args, **kwargs):
+        # If submod is wrapped with DDP, we use the `no_sync` context manager to
+        # avoid gradient all-reduce per microbatch
+        def forward_maybe_with_ddp(*args, **kwargs):
+            if isinstance(self.submod, DistributedDataParallel):
+                with self.submod.no_sync():  # type: ignore[operator]
+                    out_val = self.submod(*args, **kwargs)
+            else:
+                out_val = self.submod(*args, **kwargs)
+            return out_val
+
+        def backward_maybe_with_ddp(bwd_kwargs: Dict, is_last_chunk: bool):
+            if isinstance(self.submod, DistributedDataParallel):
+                if is_last_chunk:
+                    # HACK: reaching into DDP implementation details here. Is there a better way?
+                    self.submod.reducer.prepare_for_backward(  # type: ignore[union-attr, operator]
```

Comment:
It seems to me that unless we change DDP's code, it is unlikely we can do without an intrusive call here.

Reply:
@mrshenli I wonder if you have any suggestion?

Reply:
Would I be correct the problem is that DDP expects interleaving fwd and bwd pass, but here we need multiple fwd (i.e., # of micro-batches) and then multiple bwd?

I agree with @fegin that we might need to change DDP source code to create a dedicated API for this. 

One concern is that the contract of `no_sync()` is to call one pair of fwd+bwd after exiting the context. This PR is doing prepare_for_backward + bwd. `no_sync` also has impact on pre-forward logic like this one:

https://github.com/pytorch/pytorch/blob/c00dd43e43e9bdd041e44e5293c3fb8cd9e1bbf0/torch/nn/parallel/distributed.py#L1383-L1386

Reply:
@fegin I am also wondering why the code here can work.

Also, here pippy accumulate grad while DDP do the average. Is it possible that we can do grad average across all micro-batches? We got this feedback from Mosaic AI and they said TP+FSDP does not have same loss curve. In TP, we do grad accumulation as well.

Reply:
> here we need multiple fwd (i.e., # of micro-batches) and then multiple bwd?

Yes, @mrshenli, that's the use case here. 

Summarizing the comments so far, we have two questions:

Q1: is it okay to accumulate gradients across micro-batches? Or do we need gradient averaging?

Q2: if `no_sync()` is not ideal, do you know if there is a better way to compose with DDP? Thanks!

Reply:
@mrshenli  It seems that `prepare_forward()` is for statistics and log only. Although it is not ideal to lose some functions, the accuracy should not be affected? 

@kwen2501 I feel `no_sync()` is the right way to go since `no_sync()` is created for people manually doing micro-batch for large activation. Can we verify the gradient with pure DDP to further understand the case?

Reply:
@kwen2501 @fegin Could you folks help me further understand the constraints here and why we need to use prepare_for_backward internal calls? If we need to support multiple forward + multiple backward under no_sync, shall we try to make that work in today's DDP? 

My concern with prepare_for_backward, call to _find_tensors, etc is that there may be edge cases around static_graph, find_unused_parameters, etc that we miss and we should try to resolve all the issues using public supported DDP APIs.

Reply:
Could we land this version first if the accuracy is correct? This way we can unblock the testing TP + DDP. We can discuss how to do a proper change in DDP.

Reply:
sure!

Reply:
Hi @fegin @rohan-varma I added gradient equivalence test in `local_test_c10d_ddp.py`. The test is passing. 
Would you mind taking a second look? Thanks!